lib/SearchHelper.php

```diff
@@ -0,0 +1,80 @@
+<?php
+/**
+ * @copyright Copyright (c) 2018 Robin Appelman <robin@icewind.nl>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Mail;
+
+
+class SearchHelper {
+	const FLAG_MAP = [
+		'read' => ['SEEN', true],
+		'unread' => ['SEEN', false],
+		'answered' => ['ANSWERED', true],
+	];
+
+	/**
+	 * @param string $filter
+	 * @return \Horde_Imap_Client_Search_Query
+	 */
+	public function parseFilterString($filter) {
+		$query = new \Horde_Imap_Client_Search_Query();
+		if (!$filter) {
+			return $query;
+		}
+		$tokens = explode(' ', $filter);
+		$textTokens = [];
+		foreach ($tokens as $token) {
+			if (!$this->parseFilterToken($query, $token)) {
+				$textTokens[] = $token;
+			}
+		}
+		if (count($textTokens)) {
+			$query->text(implode(' ', $textTokens), false);
```

Comment:
Could you please share some insight on why you're adding the tokens as text search query here as well, e.g. when they are header text like subjects?

Reply:
`$textTokens` are all parts of the search query which were not parsed to have a special meaning, and just get used for "normal" text search

Reply:
You're right, I misread the condition :see_no_evil: